plugins/soundsourcem4a/audiosourcem4a.cpp

```diff
@@ -0,0 +1,336 @@
+    pDecoderConfig->defObjectType = LC;
+    if (!NeAACDecSetConfiguration(m_hDecoder, pDecoderConfig)) {
+        qWarning() << "Failed to configure AAC decoder!";
+        return ERR;
+    }
+
+    u_int8_t* configBuffer = NULL;
+    u_int32_t configBufferSize = 0;
+    if (!MP4GetTrackESConfiguration(m_hFile, m_trackId, &configBuffer,
+            &configBufferSize)) {
+        /* failed to get mpeg-4 audio config... this is ok.
+         * NeAACDecInit2() will simply use default values instead.
+         */
+        qWarning() << "Failed to read the MP4 audio configuration."
+                << "Continuing with default values.";
+    }
+
+    SAMPLERATE_TYPE sampleRate;
+    unsigned char channelCount;
+    if (0
+            > NeAACDecInit2(m_hDecoder, configBuffer, configBufferSize,
+                    &sampleRate, &channelCount)) {
+        free(configBuffer);
+        qWarning() << "Failed to initialize the AAC decoder!";
+        return ERR;
+    } else {
+        free(configBuffer);
+    }
+
+    setChannelCount(channelCount);
+    setFrameRate(sampleRate);
+    setFrameCount(m_maxSampleBlockId * kFramesPerSampleBlock);
+
+    // Allocate one block more than the number of sample blocks
+    // that are prefetched
+    const SampleBuffer::size_type prefetchSampleBufferSize =
+            (kNumberOfPrefetchSampleBlocks + 1)
+                    * frames2samples(kFramesPerSampleBlock);
+    m_prefetchSampleBuffer.resize(prefetchSampleBufferSize);
+
+    m_curSampleBlockId = 0;
+    m_curFrameIndex = 0;
+
+    return OK;
+}
+
+void AudioSourceM4A::close() {
+    if (m_hDecoder) {
+        NeAACDecClose(m_hDecoder);
+        m_hDecoder = NULL;
+    }
+    if (MP4_INVALID_FILE_HANDLE != m_hFile) {
+        MP4Close(m_hFile);
+        m_hFile = MP4_INVALID_FILE_HANDLE;
+    }
+    m_trackId = MP4_INVALID_TRACK_ID;
+    m_maxSampleBlockId = MP4_INVALID_SAMPLE_ID;
+    m_curSampleBlockId = MP4_INVALID_SAMPLE_ID;
+    m_inputBuffer.clear();
+    m_inputBufferOffset = 0;
+    m_inputBufferLength = 0;
+    m_curFrameIndex = 0;
+    reset();
+}
+
+bool AudioSourceM4A::isValidSampleBlockId(MP4SampleId sampleBlockId) const {
+    return (sampleBlockId >= kMinSampleBlockId)
+            && (sampleBlockId <= m_maxSampleBlockId);
+}
+
+void AudioSourceM4A::restartDecoding(MP4SampleId sampleBlockId) {
+    NeAACDecPostSeekReset(m_hDecoder, sampleBlockId);
+    m_curSampleBlockId = sampleBlockId;
+    m_curFrameIndex = (m_curSampleBlockId - kMinSampleBlockId)
+            * kFramesPerSampleBlock;
+    // discard input buffer
+    m_inputBufferOffset = 0;
+    m_inputBufferLength = 0;
+}
+
+AudioSource::diff_type AudioSourceM4A::seekSampleFrame(diff_type frameIndex) {
+    DEBUG_ASSERT(isValidFrameIndex(frameIndex));
+    if (m_curFrameIndex != frameIndex) {
+        MP4SampleId sampleBlockId = kMinSampleBlockId
+                + (frameIndex / kFramesPerSampleBlock);
+        DEBUG_ASSERT(isValidSampleBlockId(sampleBlockId));
```

Comment:
The DEBUG_ASSERTs should assist during the development to catch out-of-range errors and violation of invariants instantly. We can decide later, which of them are actually needed as runtime checks in release builds for reliable operation and replace them with explicit checks and corresponding log messages.